utils/tfw_logger.cpp

```diff
@@ -0,0 +1,214 @@
+{
+	Block *block = tfw_clickhouse->getBlock(cpu);
+	const char *p = data;
+	TfwBinLogEvent *event = (TfwBinLogEvent *)p;
+	int i;
+
+	p += sizeof(TfwBinLogEvent);
+	size -= sizeof(TfwBinLogEvent);
+
+#define INT_CASE(method, col_type, val_type)					\
+	case method:								\
+		if (len > size)							\
+			return -1;						\
+		if (TFW_MMAP_LOG_FIELD_IS_SET(event, i))			\
+			(*block)[ind]->As<col_type>()->Append(*(val_type *)p);	\
+		else								\
+			(*block)[ind]->As<col_type>()->Append(0);		\
+		break;
+
+
+	(*block)[0]->As<ColumnUInt64>()->Append(event->timestamp);
+
+	for (i = TFW_MMAP_LOG_ADDR; i < TFW_MMAP_LOG_MAX; ++i) {
+		int len, ind = i + 1;
+
+		len = tfw_mmap_log_field_len((TfwBinLogFields)i);
+
+		switch (i) {
+		INT_CASE(TFW_MMAP_LOG_ADDR, ColumnIPv6, struct in6_addr);
+		INT_CASE(TFW_MMAP_LOG_METHOD, ColumnUInt8, unsigned char);
+		INT_CASE(TFW_MMAP_LOG_VERSION, ColumnUInt8, unsigned char);
+		INT_CASE(TFW_MMAP_LOG_STATUS, ColumnUInt16, uint16_t);
+		INT_CASE(TFW_MMAP_LOG_RESP_CONT_LEN, ColumnUInt32, uint32_t);
+		INT_CASE(TFW_MMAP_LOG_RESP_TIME, ColumnUInt32, uint32_t);
+
+		case TFW_MMAP_LOG_VHOST:
+		case TFW_MMAP_LOG_URI:
+		case TFW_MMAP_LOG_REFERER:
+		case TFW_MMAP_LOG_USER_AGENT:
+			if (!TFW_MMAP_LOG_FIELD_IS_SET(event, i)) {
+				(*block)[ind]->As<ColumnString>()->Append(move(string("")));
+				break;
+			}
+			len = *((uint16_t *)p);
+			if (len + 2 > size)
+				return -1;
+			(*block)[ind]->As<ColumnString>()->Append(move(string(p + 2, len)));
+			len += 2;
+			break;
+		default:
+			cerr << "Unknown field type: " << i << endl;
+			return -1;
+		}
+
+		if (TFW_MMAP_LOG_FIELD_IS_SET(event, i)) {
+			p += len;
+			size -= len;
+		}
+	}
+
+	return p - data;
+}
+
+void
+callback(const char *data, int size, unsigned int cpu)
+{
+	TfwBinLogEvent *event;
+	const char *p = data;
+	int r;
+
+#ifdef DEBUG
+	hexdump(data, size);
+#endif /* DEBUG */
+
+	do {
+		if (size < (int)sizeof(TfwBinLogEvent))
+			return;
+
+		event = (TfwBinLogEvent *)p;
+
+		switch (event->type) {
+		case TFW_MMAP_LOG_TYPE_ACCESS:
+			r = read_access_log_event(p, size, cpu);
+			if (r < 0)
+				return;
+			size -= r;
+			p += r;
+			break;
+		case TFW_MMAP_LOG_TYPE_DROPPED:
+			p += sizeof(TfwBinLogEvent);
+			size -= sizeof(TfwBinLogEvent);
+			if ((int)sizeof(u64) > size) {
+				cerr << "Incorrect event length" << endl;
+				return;
+			}
+
+			cerr << "Dropped events: " << *(u64 *)p << endl;
+			size -= sizeof(u64);
+			p += sizeof(u64);
+			return;
+		default:
+			cerr << "Unsupported log type: " << event->type << endl;
+			return;
+		}
+	} while (size > 0);
+
+	tfw_clickhouse->commit(cpu);
+}
+
+int
+main(int argc, char* argv[])
+{
+	unsigned int cpu_cnt = sysconf(_SC_NPROCESSORS_ONLN);
+
+	if (argc != 2) {
+		cout << "Usage:" << endl;
+		cout << "\t" << argv[0] << " <host>" << endl;
+		return -EINVAL;
+	}
+
+	tfw_column_factory = new TfwColumnFactory();
+	tfw_clickhouse = new TfwClickhouse(argv[1], TABLE_NAME, cpu_cnt, make_block);
```

Comment:
So if we pass `-h` as a command line argument, then we'll be using it as the host name. Typically we use boost::program_options for parsing arguments, reuse the code from the manager daemon and cli

Reply:
Switched to boost arguments handling.